### PR TITLE
Update vertical line integral metrics

### LINKED ISSUE #1827: Revisit vertical integral metrics
Right now we compute the vertical line integral metrics from the volume metrics. As @Zhengyu-Huang realized, these are not exact in simple cases.

We may want to revisit these calculations and compute them directly from the radial grid and not the volume metrics.

One problem right now is that if you integrate the constant function it will not be exact on the sphere (e.g., scaling of outer minus inner radius). It will be accurate (convergent under mesh refinement), but not exact.

__Originally posted by @jkozdon in https://github.com/CliMA/ClimateMachine.jl/pull/1810#issuecomment-739303044__

### PR DESCRIPTION
**Pull request by jkozdon**
Previously the vertical line integrals metrics were computed from the 3-D metrics. This means that they are not exact in simple cases (spherical mesh), since we use curl invariant metrics. This updates the vertical integral metrics to compute them directly from the point distributions in the vertical.

This update ensures that vertical integrals of `1` will be exact on the sphere and the value of `JcV` is half the vertical element size for `N = 0`

Also adds new tests for exactness of vertical integral on the sphere and that `N = 0` gives correct value for element size.

Additional updates will be needed for the `N = 0` case, since the integral values will be at the element faces (not the cell center).

Note: That previously `bottom1` values in the boundary treatment lead to invalid array access for `N=0`, and these fixes are included in this PR (fix needed for CI purposes)

Co-authored-by: Daniel Z. Huang <hzyhzy2014@gmail.com>
Co-authored-by: Valeria Barra <valeriabarra21@gmail.com

closes #1827

### CODE REVIEW

**Comment by jkozdon** [comment]
This is an alternative approach to the issues addressed by #1810

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
@blallen @glwagner @christophernhill @jm-c:

This tweeks the metrics a bit, which results in some slight changes to ocean CI (see this commit: 582414c0bc63b09b275c5eb6e2e755e9e18316f4)

Two things:

1. Can you check the ocean results to make sure they are acceptable?
2. I am having trouble finding numbers that pass both the CPU and GPU CI. Thoughts? (https://buildkite.com/clima/climatemachine-ci/builds/839)

**Comment by valeriabarra** [comment]
> 
> Co-authored-by: Daniel Z. Huang [hzyhzy2014@gmail.com](mailto:hzyhzy2014@gmail.com)
> Co-authored-by: Valeria Barra [39932030+valeriabarra@users.noreply.github.com](mailto:39932030+valeriabarra@users.noreply.github.com)
> 
> closes #1827

Thanks for the co-authored tag @jkozdon ! It was really not necessary, since you had helped me with those changes. Not sure why I have got that weird email though. My GH email is valeriabarra21@gmail.com

**Comment by simonbyrne** [comment]
> Thanks for the co-authored tag @jkozdon ! It was really not necessary, since you had helped me with those changes. Not sure why I have got that weird email though. My GH email is different

GitHub masks them out. See https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address

**Comment by valeriabarra** [comment]
> > Thanks for the co-authored tag @jkozdon ! It was really not necessary, since you had helped me with those changes. Not sure why I have got that weird email though. My GH email is different
> 
> GitHub masks them out. See https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address

Oh, I see! Yeah I was wondering if that was it. Makes sense they would do it!
Thanks!

**Comment by jkozdon** [comment]
> Thanks for the co-authored tag @jkozdon

You helped. I like to give credit. If you want me to remove I can.

**Comment by valeriabarra** [comment]
> > Thanks for the co-authored tag @jkozdon
> 
> You helped. I like to give credit. If you want me to remove I can.

No, no it is absolutely fine :)

File: src/Numerics/Mesh/Grids.jl [hunk 1]
```diff
@@ -935,7 +918,7 @@ function indefinite_integral_interpolation_matrix(r, ω)
 
     I∫ = similar(r, Nq, Nq)
     # first value is zero
-    I∫[1, :] .= 0
+    I∫[1, :] .= Nq == ω[1] ? 2 : 0
```
Comment (valeriabarra) [inline src/Numerics/Mesh/Grids.jl:921]: I think in #1833 this is

```suggestion
    I∫[1, :] .= Nq == 1 ? ω[1] : 0
```

Comment (jkozdon) [reply src/Numerics/Mesh/Grids.jl]: Thanks.

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
> @blallen @glwagner @christophernhill @jm-c:
> 
> This tweeks the metrics a bit, which results in some slight changes to ocean CI (see this commit: [582414c](https://github.com/CliMA/ClimateMachine.jl/commit/582414c0bc63b09b275c5eb6e2e755e9e18316f4))
> 
> Two things:
> 
> 1. Can you check the ocean results to make sure they are acceptable?
> 2. I am having trouble finding numbers that pass both the CPU and GPU CI. Thoughts? (https://buildkite.com/clima/climatemachine-ci/builds/839)

I just forced push to the branch. Probably easiest to see changes in the PR diff view: https://github.com/CliMA/ClimateMachine.jl/pull/1829/files

Files I changed are:

- `test/Ocean/HydrostaticBoussinesq/test_windstress_short.jl`
- `test/Ocean/refvals/hydrostatic_spindown_refvals.jl`
- `test/Ocean/refvals/test_windstress_refvals.jl`

**Comment by blallen** [comment]
> > @blallen @glwagner @christophernhill @jm-c:
> > This tweeks the metrics a bit, which results in some slight changes to ocean CI (see this commit: [582414c](https://github.com/CliMA/ClimateMachine.jl/commit/582414c0bc63b09b275c5eb6e2e755e9e18316f4))
> > Two things:
> > 
> > 1. Can you check the ocean results to make sure they are acceptable?
> > 2. I am having trouble finding numbers that pass both the CPU and GPU CI. Thoughts? (https://buildkite.com/clima/climatemachine-ci/builds/839)
> 
> I just forced push to the branch. Probably easiest to see changes in the PR diff view: https://github.com/CliMA/ClimateMachine.jl/pull/1829/files
> 
> Files I changed are:
> 
> * `test/Ocean/HydrostaticBoussinesq/test_windstress_short.jl`
> * `test/Ocean/refvals/hydrostatic_spindown_refvals.jl`
> * `test/Ocean/refvals/test_windstress_refvals.jl`

The spindown values look fine and are still converging to the analytical solution, so it's probably fine. @christophernhill @jm-c can one of you check the barotropic gyre output (test_windstress_short.jl)?

**Comment by christophernhill** [comment]
> > @blallen @glwagner @christophernhill @jm-c:
> > This tweeks the metrics a bit, which results in some slight changes to ocean CI (see this commit: [582414c](https://github.com/CliMA/ClimateMachine.jl/commit/582414c0bc63b09b275c5eb6e2e755e9e18316f4))
> > Two things:
> > 
> > 1. Can you check the ocean results to make sure they are acceptable?
> > 2. I am having trouble finding numbers that pass both the CPU and GPU CI. Thoughts? (https://buildkite.com/clima/climatemachine-ci/builds/839)
> 
> I just forced push to the branch. Probably easiest to see changes in the PR diff view: https://github.com/CliMA/ClimateMachine.jl/pull/1829/files
> 
> Files I changed are:
> 
> * `test/Ocean/HydrostaticBoussinesq/test_windstress_short.jl`
> * `test/Ocean/refvals/hydrostatic_spindown_refvals.jl`
> * `test/Ocean/refvals/test_windstress_refvals.jl`

@jkozdon I took a look. The changes in reference numbers don't look alarming to me - we don't expect to match to every digit with a change like this. The things I checked do seem to match to ~11 digits, except for some means which are epsilon close to zero so its reasonable to change all the digits.   

@jm-c may have some extra comments, but these look OK to me.

**Comment by jkozdon** [comment]
Thanks @christophernhill and @blallen

**Comment by jkozdon** [comment]
@Zhengyu-Huang You good with this?

**Review by thomasgibson — APPROVED** [review]
LGTM! Thanks for taking care of this

**Pull request merged**